drivers/platform/mbed/mbed_timer.cpp

```diff
@@ -0,0 +1,191 @@
+/************************** Functions Definitions ****************************/
+/*****************************************************************************/
+
+/**
+ * @brief Initialize the timer peripheral.
+ * @param desc[in, out] - The Timer descriptor.
+ * @param init_param[in] - Structure that contains Timer Initialization Parameters.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t timer_init(struct timer_desc **desc,
+		   struct timer_init_param *param)
+{
+	struct timer_desc *timer_desc; 		// Pointer to timer descriptor
+	struct mbed_timer_desc *mbed_timer_desc;	// Pointer to mbed timer descriptor
+	mbed::Timer *timer;					// Pointer to new Timer instance
+
+	if (!desc || !param)
+		return -EINVAL;
+
+	timer_desc = (struct timer_desc *)calloc(1, sizeof(*timer_desc));
+	if (!timer_desc)
+		return -ENOMEM;
+
+	mbed_timer_desc = (struct mbed_timer_desc *)calloc(1, sizeof(*mbed_timer_desc));
+	if (!mbed_timer_desc)
+		goto err_mbed_timer_desc;
+
+	timer = new Timer();
+	if (!timer)
+		goto err_timer;
+
+	mbed_timer_desc->timer = timer;
+	timer_desc->extra = mbed_timer_desc;
+
+	*desc = timer_desc;
+	return SUCCESS;
+
+err_timer:
+	free(mbed_timer_desc);
+err_mbed_timer_desc:
+	free(timer_desc);
+
+	return -ENOMEM;
+}
+
+/**
+* @brief Start the Timer.
+* @param desc[in] - The Timer Descriptor.
+* @return SUCCESS in case of success, negative error code otherwise.
+*/
+int32_t timer_start(struct timer_desc *desc)
+{
+	mbed::Timer *timer;
+
+	if (!desc || !desc->extra)
+		return -EINVAL;
+
+	timer = (Timer *)(((struct mbed_timer_desc *)(desc->extra))->timer);
+	if (!timer)
+		return -ENOMEM;
+
+	timer->start();
+
+	return SUCCESS;
+}
+
+/**
+* @brief Stop the Timer.
+* @param desc[in] - The Timer Descriptor.
+* @return SUCCESS in case of success, negative error code otherwise.
+*/
+int32_t timer_stop(struct timer_desc *desc)
+{
+	mbed::Timer *timer;
+
+	if (!desc || !desc->extra)
+		return -EINVAL;
+
+	timer = (Timer *)(((struct mbed_timer_desc *)(desc->extra))->timer);
+	if (!timer)
+		return -ENOMEM;
+
+	timer->stop();
+
+	return SUCCESS;
+}
+
+/**
+* @brief Release all the resources allocated by Timer.
+* @param desc[in] - The Timer Descriptor.
+* @return SUCCESS in case of success, negative error code otherwise.
+*/
+int32_t timer_remove(struct timer_desc *desc)
+{
+	if (!desc || !desc->extra)
+		return -EINVAL;
+
+	/* Free the timer object */
+	if ((Timer *)(((struct mbed_timer_desc *)(desc->extra))->timer))
+		delete((Timer *)(((struct mbed_timer_desc *)(desc->extra))->timer));
+
+	/* Free the Timer descriptors */
+	free(desc->extra);
+	free(desc);
+
+	return SUCCESS;
+}
+
+/**
+* @brief Get the elapsed time in nanoseconds.
+* @param desc[in] - The Timer descriptor.
+* @param elapsed_time[in, out] - Pointer where the elapsed time value is stored.
+* @return SUCCESS in case of success, negative error code otherwise.
+*/
+int32_t get_elapsed_time_in_nsec(struct timer_desc *desc,
```

Comment:
Is there a difference between this method and `timer_counter_get`? Could we use the existing API here?

Reply:
This method returns the elapsed time in usec however I think timer_counter_get method expects the difference between previous timer count and current timer count, which would be then used to calculate the actual elapsed time in time units based on timer frequency (*Please correct me if my understanding is wrong here).

So basically mbed does not have function that gives us the difference b/w timer counts. However, we could use timer_counter_get method and treat it as an API which returns value as elapsed time directly instead of elapsed count.
Only problem is this mbed method uses INT64 type for holding timer elapsed value and timer_counter_get uses uint32 argument for return value, so we have to typecast return value from mbed function to uint32. See below:

*elapsed_time = (uint32_t )(duration_cast<nanoseconds>(timer->elapsed_time()).count());

It may loose some resolution though and result may be incorrect. So not a very good idea to do.
Let me know what you think.

Reply:
I agree with @adrimbarean that we should stick to the existing API. I'm wondering if we can use the ticker functions directly, like `timestamp_t ticker_read(const ticker_data_t *const ticker);` from https://os.mbed.com/docs/mbed-os/v6.15/mbed-os-api-doxy/ticker__api_8h_source.html since if I am not mistaken the Ticker API is included in https://os.mbed.com/docs/mbed-os/v6.6/mbed-os-api-doxy/_timer_8h_source.html

Reply:
Hi All,
Looks like lack of mbed API that directly returns timer count in decimal, makes it difficult to make mbed timer code compatible with no-os driver timer include API's. I understand 'timer_counter_get()' method is mandatory to use mbed timer module with other no-os projects. 

Do you want me to delete this PR (that means we will not add Timer support in mbed platform drivers) Or do you want to keep the changes those are done so far (without timer_counter_get() method)?

Reply:
Hello,

Let's keep this implementation. But could you add the `get_elapsed_time_in_nsec()` in the [generic timer header](https://github.com/analogdevicesinc/no-OS/blob/master/include/no-os/timer.h)? And call it something with `timer_` at the beginning. We should avoid defining API just for one platform.

Reply:
Sure. I will update the changes.